### PR TITLE
prevector: enforce is_trivially_copyable_v

### PR DESCRIPTION
**Pull request by martinus**
prevector uses `memmove` to move around data, that means it can only be used with types that are trivially copyable. That implies that the types are trivially destructible, thus the checks for `is_trivially_destructible` are not needed.

### CODE REVIEW

File: src/prevector.h [hunk 1]
```diff
@@ -410,17 +413,9 @@ class prevector {
...
-                _size--;
-                ++p;
-            }
-        } else {
-            _size -= last - p;
-        }
-        memmove(&(*first), &(*last), endp - ((char*)(&(*last))));
+        char* endp = reinterpret_cast<char*>(end().ptr);
+        _size -= last - p;
+        memmove(first.ptr, last.ptr, endp - reinterpret_cast<char*>(last.ptr));
```
Comment (MarcoFalke) [inline src/prevector.h:418]:
```suggestion
        std::memmove(first.ptr, last.ptr, endp - reinterpret_cast<char*>(last.ptr));
```

Maybe switch to `std::memmove`, which comes with the documentation:

https://en.cppreference.com/w/cpp/string/byte/memmove

> If the objects are [potentially-overlapping](https://en.cppreference.com/w/cpp/language/object#Subobjects) or not [TriviallyCopyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable), the behavior of memmove is not specified and [may be undefined](http://stackoverflow.com/questions/29777492).

**Comment by theStack** [comment]
Light Concept ACK
(I'm not deeply familiar with templates, type traits, etc. but from a higher level perspective, removing unused code and fixing UB seems like a good idea!)

File: src/prevector.h (hunk 1 above)
Comment (MarcoFalke) [reply src/prevector.h:418]: Oh, if the changes are more expensive, then maybe do it in another commit or pull request.

**Review by jonatack — COMMENTED** [review]
ACK 1ed610d3892917240589ca55720340838154742d

**Comment by MarcoFalke** [comment]
I haven't reviewed the second commit and I wonder why no sanitizer complains about them. I presume the reason is that this is no stdlib iterator? See also: https://github.com/bitcoin/bitcoin/pull/21817#discussion_r624434789 and https://github.com/bitcoin/bitcoin/pull/21869

**Comment by martinus** [comment]
> I wonder why no sanitizer complains about them

I was wondering about that too, and thought this might be because the optimizer is just optimizing that away. After some googling though it seems this is actually allowed and no UB: 

* In C there is an explicit rule that allowes to do ` &*whatever`, even when whatever is a nullptr: https://twitter.com/shafikyaghmour/status/1185235611407417344
* But in C++ there is no such rule that allows this: https://twitter.com/shafikyaghmour/status/1185235735743352833
* That issue was then addressed a few years ago: http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#232
* and now this definition should say that it's actually allowed: https://eel.is/c++draft/expr.unary.op#1

TL;DR: I was most likely wrong and it is fine to do `&*end()`

File: src/prevector.h (hunk 1 above)
Comment (jonatack) [reply src/prevector.h]: Does UB section of the documentation in https://en.cppreference.com/w/cpp/string/byte/memmove apply to https://en.cppreference.com/w/c/string/byte/memmove?

Comment (MarcoFalke) [reply src/prevector.h]: I am not familiar with C, but maybe every class is trivially copyable in C?

**Comment by jonatack** [comment]
(also, update the pull description, e.g. remove the second paragraph pertaining to the dropped commit)

File: src/prevector.h (hunk 1 above)
Comment (jonatack) [reply src/prevector.h]: Right (interesting [historic context](https://www.stroustrup.com/bs_faq.html#invention) here). So IIUC the premise in the pull description only applies to the C++ memmove and it may make sense to change to it at the same time, or not at all?

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/24962/commits/11e79084845a78e2421ea3abafe0de5a54ca2bde

File: src/prevector.h (hunk 1 above)
Comment (martinus) [reply src/prevector.h]: > the premise in the pull description only applies to the C++ memmove and it may make sense to change to it at the same time (or not do this change at all).

I still think it is necessary to do the change, even when we stay with `memmove`. The current implementation just does not work with complex objects. E.g. here is a small example:

```cpp
BOOST_AUTO_TEST_CASE(Erase) {
    prevector<8, std::string> pv;
    pv.push_back("a");
    pv.push_back("b");
    pv.erase(pv.begin());
    pv.push_back("X");
    std::cout << pv.front() << std::endl;
}
```

One would expect that this prints "b", as "a" was deleted and "b" should now be the first element. But l get this output:

```
Running 1 test case...
X
free(): invalid pointer
unknown location(0): fatal error: in "prevector_tests/Erase": signal: SIGABRT (application abort requested)
```

So suddenly the first element is not "b" but "X". This is due to to small string optimization, the strings point to member data, and this is simply copied. Complex objects have to be moved for this to work.

here is another motivating example with `std::swap`:

```cpp
BOOST_AUTO_TEST_CASE(NonTrivial) {
    prevector<8, std::string> a;
    prevector<8, std::string> b;
    a.push_back("a");
    b.push_back("b");

    std::cout << "a=" << a.front() << ", b=" << b.front() << std::endl;
    std::swap(a, b);
    std::cout << "a=" << a.front() << ", b=" << b.front() << std::endl;
}
```
Running that test produces this output:

```
Running 1 test case...
a=a, b=b
a=a, b=b
free(): invalid pointer
unknown location(0): fatal error: in "prevector_tests/NonTrivial": signal: SIGABRT (application abort requested)
```

So swap doesn't seem to do anything for these objects, and destructing will cause a SIGABRT. 

I think we should either restrict prevector to only work with supported types, or to correctly implement it for nontrivial types. Fixing this is quite a bit more work though, and would need a lot more testing.

File: src/prevector.h [hunk 2]
```diff
@@ -35,6 +35,8 @@
  */
 template<unsigned int N, typename T, typename Size = uint32_t, typename Diff = int32_t>
 class prevector {
+    static_assert(std::is_trivially_copyable_v<T>);
```
Comment (MarcoFalke) [inline src/prevector.h:38]: Out of interest, I am wondering to what extend trivially copyable is related to trivially destructible?

Comment (martinus) [reply src/prevector.h:38]: Per the standard a class is only trivially copyable when it also has a trivial destructor: https://eel.is/c++draft/class#prop-1.3

**Review by maflcko — APPROVED** [review]
review ACK 11e79084845a78e2421ea3abafe0de5a54ca2bde 🏯

I checked that the code is unused at run-time. Also, in light of silent merge
conflicts, the added static_assert should ensure that any code that tried to
use it fails to compile. According to https://eel.is/c++draft/class#prop-1.3
is_trivially_copyable_v is false if is_trivially_destructible_v is false.

I have not verified that this is actually UB, but the provided example seems to
imply so. Regardless, removing code because it is unused seems good enough.

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 11e79084845a78e2421ea3abafe0de5a54ca2bde 🏯

I checked that the code is unused at run-time. Also, in light of silent merge
conflicts, the added static_assert should ensure that any code that tried to
use it fails to compile. According to https://eel.is/c++draft/class#prop-1.3
is_trivially_copyable_v is false if is_trivially_destructible_v is false.

I have not verified that this is actually UB, but the provided example seems to
imply so. Regardless, removing code because it is unused seems good enough.
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhd7gv/X+b9gxeWmyrk43wUYLfe9bXgwHPM3JRI068e7fVFeml8flhkQIYaY4eP
Km+G8XGwEf0TaTY2Y1ExmSxsBE3EKF9sYRLLga/FOeBamKpQwBHefNOQyPEdP9wr
KQSYv7mB27U6xt8tFv+p1i9Hr4LWQVEEWX2dYCl8qio+kObOQQFI4xhManXnp06d
+8Vp+7IbS2H9jLIPK+1byiozLWaT8pr15PRRgQL1XuvlBwEdoBmF/usMP/tV4MaC
Ns8CXIXB5guiA96HDSsK0dsDzQgSqrX8ZrqSP0007lkW4kOW/TqsxMQhtTbEppOO
BadS8y0QIBcrOl2WGICJlfSqm3BKb7/YCbR+jdAaS5LHT5YxM6/k4a0KjPt1ijpM
VibEmNkimCin3cpVbxAouptTdP1KyNoeTCrDU1yGM3o/tMAy0axjPhos9Iv2avvn
xtBGzMFe6XzRQR25mvXBsEd/v7Dbe/zBJPLRFGFr7NhR+2XBSq5B7ncaj9PwK5uy
r4uqloDb
=xRUi
-----END PGP SIGNATURE-----
```


</details>

**Comment by Meru852** [comment]
> review ACK [11e7908](https://github.com/bitcoin/bitcoin/commit/11e79084845a78e2421ea3abafe0de5a54ca2bde) 🏯
> 
> I checked that the code is unused at run-time. Also, in light of silent merge conflicts, the added static_assert should ensure that any code that tried to use it fails to compile. According to https://eel.is/c++draft/class#prop-1.3 is_trivially_copyable_v is false if is_trivially_destructible_v is false.
> 
> I have not verified that this is actually UB, but the provided example seems to imply so. Regardless, removing code because it is unused seems good enough.
> 
> Show signature

hello Marco. do you check your telegrams? some one need your reply by tele.

**Review by ajtowns — APPROVED** [review]
ACK 11e79084845a78e2421ea3abafe0de5a54ca2bde -- code review only

File: src/prevector.h (hunk 1 above)
Comment (fanquake) [reply src/prevector.h]: Somewhat followed up in #25172.
